File: src/java/mp/ShMem.java
package mp;

import java.util.*;
import java.net.*;
import java.io.*;

import org.apache.commons.lang3.tuple.*;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.node.ArrayNode;
import org.codehaus.jackson.node.ObjectNode;


//
// Use this class to manage forks. Automatically keeps track of the 
// reference state, and merges with the parent state. The methods in 
// this class are NOT thread-safe. 
//
public class ShMem {
	
	// Address book populated from disk. 
	public static Map<Integer, Pair<String, Integer>> addresses;
	public static ObjectMapper mapper;

	static {
		Properties prop = new Properties();
		String config_path = "mp.properties";
		addresses = new HashMap<Integer, Pair<String, Integer>>();	
		mapper = new ObjectMapper();
		try {
			FileInputStream fi = new FileInputStream(config_path);
			prop.load(fi);
			String input_file = prop.getProperty("addresses");
			ShMem.populate_addresses(input_file);
		}
		catch (Exception e) {
			System.out.println("Couldn't read addresses properly!");
			e.printStackTrace();
			System.exit(-1);
		}
	}

	
	public  static ShMemObject s_state;	// Externally visible state.
	private static DeltaStore deltas_;			// Repository of received deltas. 
	private static ShMemReleaser releaser_; 	// Use to release deltas. 
	private static ShMemAcquirer acquirer_;
	
	// We need to keep the last time we released to extract the appropriate
	// deltas from state_ while releasing. 
	private static Map<Integer, int[]> m_last_sync;  
	
	public static void InitTest(int node_id) {
		VectorTimestamp.Init(addresses.size(),  node_id);
		VectorTimestamp.CreateDefault();
		VectorTimestamp.s_zero = VectorTimestamp.CreateZero();
		m_last_sync = new HashMap<Integer, int[]>();
		
		for (int i = 0; i < addresses.size(); ++i)
			m_last_sync.put(i,  VectorTimestamp.CreateZero());
		
		ShMemObject.s_now = VectorTimestamp.Copy(VectorTimestamp.s_zero); 
		ShMemObject.cur_node_ = node_id;
		
		s_state = new ShMemObject();
		
		deltas_ = new DeltaStore();
		// acquirer_ = new ShMemAcquirer(node_id, deltas_);
		// releaser_ = new ShMemReleaser(node_id);
	}
	
	public static void Init(int node_id) {
		VectorTimestamp.Init(addresses.size(),  node_id);
		VectorTimestamp.CreateDefault();
		VectorTimestamp.s_zero = VectorTimestamp.CreateZero();
		m_last_sync = new HashMap<Integer, int[]>();
		
		for (int i = 0; i < addresses.size(); ++i)
			m_last_sync.put(i,  VectorTimestamp.CreateZero());
		
		ShMemObject.s_now = VectorTimestamp.Copy(VectorTimestamp.s_zero); 
		ShMemObject.cur_node_ = node_id;
		
		s_state = new ShMemObject();
		
		deltas_ = new DeltaStore();
		acquirer_ = new ShMemAcquirer(node_id, deltas_);
		releaser_ = new ShMemReleaser(node_id);
	}
	
	public static void Start() {
		VectorTimestamp.IncrementLocal(ShMemObject.s_now);
	}
	
	// Called from the application. 
	// Acquire state from another process. 
	// XXX: Automatically blocks the calling thread if
	// the state isn't yet available. 
	public static void Acquire(int from) throws ShMemObject.MergeException {
		int[] since = m_last_sync.get(from);
		
		// - Get the appropriate delta.
		// - Merge it into state_. 
		// - Change now_ appropriately.
		// - Update last_sync_.
		ObjectNode delta = deltas_.Pop(from);
		s_state.merge(delta.get("value"));
		VectorTimestamp.UnionWithSerialized(ShMemObject.s_now,  (ArrayNode)(delta.get("time")));
		VectorTimestamp.UnionWithSerialized(since,  (ArrayNode)delta.get("time"));
		VectorTimestamp.IncrementLocal(ShMemObject.s_now);
	}
	
	private static ArrayNode CopyArray(ArrayNode node) {
		
		// Create an empty ArrayNode to return, iterate through the 
		// values in the ArrayNode.
		ArrayNode ret = mapper.createArrayNode();
		int num_vals = node.size();
		for (int i = 0; i < num_vals; ++i) {
			
			// We only recurse if the value is an ObjectNode or an ArrayNode. 
			JsonNode value = node.get(i);
			if (value instanceof ObjectNode) {
				value = CopyObject((ObjectNode)value);
			}
			else if (value instanceof ArrayNode) {
				value = CopyArray((ArrayNode)value);
			}
			
			// Add the copied value to the array node we wish to return. 
			ret.add(value);
		}
		return ret;
	}
	
	private static ObjectNode CopyObject(ObjectNode node) {
		
		// Create an object node to return and iterate through the key-value
		// pairs of this object. 
		ObjectNode ret = mapper.createObjectNode();
		Iterator<Map.Entry<String, JsonNode>> iter = node.getFields();
		while (iter.hasNext()) {
			Map.Entry<String, JsonNode> kvp = iter.next();
			
			// We only need to recurse if the value is an ObjectNode or an ArrayNode
			// because nodes of simple types are immutable. 
			JsonNode to_add = kvp.getValue();
			if (to_add instanceof ObjectNode) {
				to_add = CopyObject((ObjectNode)kvp.getValue());
			}
			else if(to_add instanceof ArrayNode) {
				to_add = CopyArray((ArrayNode)to_add);
			}
			
			// Add the key-value pair to the object to return.
			ret.put(kvp.getKey(), to_add);
		}
		return ret;
	}
	
	public static ObjectNode AcquirePlain(int from) {
		return deltas_.Pop(from);
	}
	
	// Called by regular message passing applications. 
	public static void ReleasePlain(ObjectNode to_send, int to) {
		
		// Make a deep copy of the object we want to release and put it in 
		// the send queue. 
		ObjectNode deep_copy = CopyObject(to_send);	
		releaser_.Release(to,  deep_copy);
	}
	
	// Called from the application. Release state to another process. 
	// Releasing does not block, happens asynchronously in the background. 
	public static void Release(int to) {
		
		// Get the timestamp of the last release. 
		int[] last_sync = m_last_sync.get(to);
		
		// Extract the appropriate delta from state_ (all modifications to
		// state since the last time we released to the process we're releasing
		// to).
		ObjectNode to_release = mapper.createObjectNode();
		to_release.put("time",  VectorTimestamp.toArrayNode(ShMemObject.s_now));
		ObjectNode cur_delta = null;
		cur_delta = ShMemObject.get_diff_tree(s_state, last_sync);
		assert(cur_delta != null);						// Make sure we have a valid delta. 
		to_release.put("value",  cur_delta);
		VectorTimestamp.CopyFromTo(ShMemObject.s_now,  last_sync);
		releaser_.Release(to,  to_release);				// Release (asynchronously). 
		VectorTimestamp.IncrementLocal(ShMemObject.s_now);								// Increment time.
		
	}

	
	
	//
	// Populate the address book in memory with the addresses from disk. Boring
	// file parsing code. 
	// 
	private static void populate_addresses(String input_file) {
		
		FileReader reader = null;
		try {
			reader = new FileReader(input_file);
		}
		catch(FileNotFoundException e) {
			e.printStackTrace();
			System.exit(-1);
		}
		
    	BufferedReader file = new BufferedReader(reader);
    	int node_num = 0;
    	while (true) {
    		
    		String file_line = null;
    		try {
    			file_line = file.readLine();
    		}
    		catch(IOException e) {
    			e.printStackTrace();
    			System.exit(-1);
    		}
    		
    		if (file_line == null) {
    			break;
    		}
    		
    		String parts[] = file_line.split(" ");
    		
    		int port = -1;
    		
    		try {
    			
    			port = Integer.parseInt(parts[1]);
    		}
    		catch(NumberFormatException e) {
    			e.printStackTrace();
    			System.exit(-1);
    		}
    		
    		if (node_num != -1 && port != -1) {
    			addresses.put(node_num, Pair.of(parts[0],  port));
    		}
    		
    		node_num += 1;
    	}
	}

}

File: src/python/mp/ShMem.py
from ShMemObject import *
from ShMemAcquirer import *
from ShMemReleaser import *
from timestamp_util import *
from Queue import *
import copy


# Static fields:
# s_addresses		-- Pairs of (IP, port) on which remote procs live.
# s_state		-- This node's ShMemObject root. 
# s_index		-- The global id of this node (int)
# s_send_queue		-- Contains state to be asynchronously sent. 
# s_receive_queues	-- Dictionary of queues of received state.
# s_last_sync		-- Dictionary of the last sync time with other nodes. 
# s_acquirer		-- Instance of ShMemAcquirer for this node. 
# s_releaser 		-- Instance of ShMemReleaser for this node. 
class ShMem:    

    # Populate the "address book" with the addresses (IP, port) pairs of 
    # nodes in the system. 
    @staticmethod
    def populate_addresses(input_file):
        ret = {}
        counter = 0
        for line in input_file:
            parts = line.split(' ')
            listen_port = int(parts[1])
            ret[counter] = (parts[0], listen_port)
            counter += 1
        return ret

    # Initialize s_state. All nodes inserted into s_state after this call 
    # but before the call to start are not shipped to other nodes. We do this
    # so that all nodes can potentially start with the same initial state. 
    @staticmethod
    def init(input_file, my_id):
        ShMem.s_index = my_id
        
        # Get the address of each node. 
        ShMem.s_addresses = ShMem.populate_addresses(input_file)
        Timestamp.init(len(ShMem.s_addresses), my_id)
        ShMemObject.s_now = Timestamp.CreateZero()
        ShMem.s_state = ShMemObject()

        # Initialize a single send queue which we'll use to communicate with
        # an ShMemReleaser thread. 
        ShMem.s_send_queue = Queue()

        # Initialize a set of receive queues which the ShMemAcquirer thread will
        # use to put recieved state.
        ShMem.s_receive_queues = {}
        for i in range(0, len(ShMem.s_addresses)):
            if i != my_id:
                ShMem.s_receive_queues[i] = Queue()
        
    # Once this function is called, all changes made to s_state are tracked
    # and sent as part of this diffing process. 
    @staticmethod
    def start(test=True):
        ip, port = ShMem.s_addresses[ShMem.s_index]
        total_nodes = len(ShMem.s_addresses)

        if test:
            print 'here'
            ShMem.s_acquirer = ShMemAcquirer(port, 
                                             ShMem.s_index, 
                                             ShMem.s_receive_queues)
            ShMem.s_releaser = ShMemReleaser(ShMem.s_index, ShMem.s_send_queue,
                                             ShMem.s_addresses)

        # Keep a dictionary of the last time we synced with a remote node. 
        ShMem.s_last_sync = {}
        for i in range(0, total_nodes):
            ShMem.s_last_sync[i] = Timestamp.CreateZero()
        
        Timestamp.LocalIncrement(ShMemObject.s_now)
        
    # Acquire a python dictionary from a remote node. 
    @staticmethod
    def AcquirePlain(node):
        return ShMem.s_receive_queues[node].get(True)

    # Merge s_state with the state we've received from a remote node. 
    @staticmethod
    def Acquire(node):
        delta = ShMem.s_receive_queues[node].get(True)
        #print '\nBegin acquired! <<<<<<<<<\n'
        #print ShMem.rec_dict_keys(delta)
        #print '\nEnd acquired! >>>>>>>>>>>>\n'
        ShMem.s_state.merge(delta['value'])
        Timestamp.Union(ShMemObject.s_now, delta['time'])
        Timestamp.Union(ShMem.s_last_sync[node], delta['time'])
        Timestamp.LocalIncrement(ShMemObject.s_now)

    @staticmethod
    def rec_dict_keys(d):
        ret = {}
        for key,value in d.items():
            if key == 'shmem_timestamp':
                ret[key] = value
                continue

            if isinstance(value, dict):
                ret[key] = ShMem.rec_dict_keys(value)
            else:
                ret[key] = ""
        return ret
            
    
    # Use this static method to release an object to a remote node. 
    @staticmethod
    def ReleasePlain(to_send, node):
        
        # Create a deep copy of the node to send. 
        to_send_copy = copy.deepcopy(to_send)
        ShMem.s_releaser.send(node, to_send)
    
    # Release our diffs to another node. Put it in the send queue for the
    # ShMemReleaser thread to get to. 
    @staticmethod
    def Release(node):
        to_send = {}
        to_send['time'] = Timestamp.CreateCopy(ShMemObject.s_now)
        to_send['value'] = ShMem.s_state.get_diffs(ShMem.s_last_sync[node])

        #print '\nBegin release! <<<<<<<<<\n'
        #print ShMem.rec_dict_keys(to_send['value'])
        #print '\nEnd release! >>>>>>>>>>>>\n'


        Timestamp.Copy(ShMemObject.s_now, ShMem.s_last_sync[node])
        ShMem.s_releaser.send(node, to_send)
        Timestamp.LocalIncrement(ShMemObject.s_now)
